File: core/data/tests/externally_tagged_enum_edge_cases/output.scala
package com.agilebits.onepassword

import io.circe._
import io.circe.generic.semiauto._
import io.circe.syntax._

case class NestedStruct(
	nested_field: String
)

object NestedStruct {
	implicit val decoder: Decoder[NestedStruct] = deriveDecoder
	implicit val encoder: Encoder[NestedStruct] = deriveEncoder
}

/** Tests various edge cases */
sealed trait EdgeCaseEnum

object EdgeCaseEnum {
	/** Empty variant */
	case object Empty extends EdgeCaseEnum
	
	/** Optional string */
	case class OptionalString(value: Option[String]) extends EdgeCaseEnum
	
	/** Optional nested struct */
	case class OptionalNested(value: Option[NestedStruct]) extends EdgeCaseEnum
	
	/** Array of values */
	case class Array(value: List[Int]) extends EdgeCaseEnum
	
	/** Array of structs */
	case class StructArray(value: List[NestedStruct]) extends EdgeCaseEnum
	
	/** Complex nested structure */
	case class Complex(
		a: String,
		b: Option[Int],
		c: List[NestedStruct],
		d: Option[List[String]]
	) extends EdgeCaseEnum
	
	implicit val encoder: Encoder[EdgeCaseEnum] = Encoder.instance {
		case Empty => Json.fromString("Empty")
		case OptionalString(value) => Json.obj("OptionalString" -> value.asJson)
		case OptionalNested(value) => Json.obj("OptionalNested" -> value.asJson)
		case Array(value) => Json.obj("Array" -> value.asJson)
		case StructArray(value) => Json.obj("StructArray" -> value.asJson)
		case Complex(a, b, c, d) => Json.obj(
			"Complex" -> Json.obj(
				"a" -> a.asJson,
				"b" -> b.asJson,
				"c" -> c.asJson,
				"d" -> d.asJson
			)
		)
	}
	
	implicit val decoder: Decoder[EdgeCaseEnum] = new Decoder[EdgeCaseEnum] {
		def apply(c: HCursor): Decoder.Result[EdgeCaseEnum] = {
			if (c.value.isString && c.value.asString.contains("Empty")) {
				Right(Empty)
			} else {
				val keys = c.keys.getOrElse(Nil)
				if (keys.isEmpty) {
					Left(DecodingFailure("Expected object with a single key", c.history))
				} else {
					val key = keys.head
					key match {
						case "OptionalString" => c.downField("OptionalString").as[Option[String]].map(OptionalString)
						case "OptionalNested" => c.downField("OptionalNested").as[Option[NestedStruct]].map(OptionalNested)
						case "Array" => c.downField("Array").as[List[Int]].map(Array)
						case "StructArray" => c.downField("StructArray").as[List[NestedStruct]].map(StructArray)
						case "Complex" => 
							for {
								a <- c.downField("Complex").downField("a").as[String]
								b <- c.downField("Complex").downField("b").as[Option[Int]]
								c <- c.downField("Complex").downField("c").as[List[NestedStruct]]
								d <- c.downField("Complex").downField("d").as[Option[List[String]]]
							} yield Complex(a, b, c, d)
						case _ => Left(DecodingFailure(s"Unknown key: $key", c.history))
					}
				}
			}
		}
	}
}

File: core/data/tests/externally_tagged_enum_edge_cases/output.kt
package com.agilebits.onepassword.colorsModule

import kotlinx.serialization.*
import kotlinx.serialization.json.*
import kotlinx.serialization.descriptors.*
import kotlinx.serialization.encoding.*

@Serializable
data class NestedStruct(
	val nested_field: String
)

/** Tests various edge cases */
@Serializable(EdgeCaseEnumSerializer::class)
sealed class EdgeCaseEnum {
	/** Empty variant */
	@Serializable
	object Empty : EdgeCaseEnum()
	
	/** Optional string */
	@Serializable
	data class OptionalString(val value: String?) : EdgeCaseEnum()
	
	/** Optional nested struct */
	@Serializable
	data class OptionalNested(val value: NestedStruct?) : EdgeCaseEnum()
	
	/** Array of values */
	@Serializable
	data class Array(val value: List<Int>) : EdgeCaseEnum()
	
	/** Array of structs */
	@Serializable
	data class StructArray(val value: List<NestedStruct>) : EdgeCaseEnum()
	
	/** Complex nested structure */
	@Serializable
	data class Complex(
		val a: String,
		val b: Int?,
		val c: List<NestedStruct>,
		val d: List<String>?
	) : EdgeCaseEnum()
}

class EdgeCaseEnumSerializer : JsonContentPolymorphicSerializer<EdgeCaseEnum>(EdgeCaseEnum::class) {
	override fun selectDeserializer(element: JsonElement): DeserializationStrategy<out EdgeCaseEnum> {
		return when {
			element is JsonPrimitive && element.isString && element.content == "Empty" -> EdgeCaseEnum.Empty.serializer()
			element is JsonObject && "OptionalString" in element -> EdgeCaseEnum.OptionalString.serializer()
			element is JsonObject && "OptionalNested" in element -> EdgeCaseEnum.OptionalNested.serializer()
			element is JsonObject && "Array" in element -> EdgeCaseEnum.Array.serializer()
			element is JsonObject && "StructArray" in element -> EdgeCaseEnum.StructArray.serializer()
			element is JsonObject && "Complex" in element -> EdgeCaseEnum.Complex.serializer()
			else -> throw SerializationException("Unknown element: $element")
		}
	}
}